### ISSUE TITLE
BigGAN: flood loss ablation (pausing, or gradient ascent?)

### ISSUE THREAD

**Issue by gwern** [tensorfork/tensorfork#22]
[Ishida et al 2020](https://arxiv.org/abs/2002.08709 "Do We Need Zero Training Loss After Achieving Zero Training Error?") introduces 'flood loss', a new gradient descent trick: if the loss falls below a certain magnitude and the model has gotten 'too good', no gradient descent is done, but instead, gradient *ascent* is done - the model parameters are actively updated to do *worse* on the loss. The justification for Ishida is that the model can 'bounce around' a wide flat fitness landscape and eventually stop at a model which generalizes better rather than simply converging on the nearest point in model-space with a near-zero loss. If it finds a narrow local optima, the flood loss will eventually, by taking gradient ascent steps, bounce it out of the narrow optima; only wide flat minima will keep it around taking random walks.

We immediately imported this into BigGAN with settings like d_flood = 0.2 and g_flood = 0.5, because of BigGAN's serious stability issues, and... this one-weird-trick-that-NNs-hate seems to have worked? Even BigGANs which have been *severely* thrown out of whack (by having new datasets dropped in or half-baked SimCLR losses going haywire) have sorta continued training.

However, the original Ishida justification doesn't really make sense for GANs. They are two-player games, the fitness landscape changes on its own, there's no need to 'flood' the landscape with random steps. Nor are our GANs anywhere near wide flat minima early on in training when we see flood loss invoked most often during runs (with their characteristic 'bounces' on the loss curves in Tensorboard).

Do we need flood loss? Flood loss can be seen for GANs as doing two things:

1. freezing training of a G or D which has gotten 'too good'
2. actively worsening said G/D

Both of these steps allow the other half to 'catch up', but it's not obvious we need \#2. It may be that it's adequate to do \#1, and simply give the other half a free gradient descent step.

If so, that may be more efficient (since presumably the worsening in \#2 must then be immediately undone) and simpler to implement. (The improvement would be minor, however, and this is mostly a matter of intellectual curiosity and wondering how exactly flood loss is helping us; I can't immediately see any important implications of the gradient ascent step being unnecessary or unnecessary.)

And if that is true, that's interesting because pausing the training of a runaway G/D is an obvious trick which I am pretty certain lots of early GAN researchers back in 2014--2017 tried out to stabilize their GANs, and needless to say, it did not succeed---so why would it work so well for *BigGAN*?

As an ablation test, it's easy. Do a run where the flood loss causes the training loop to either skip doing a step for that model, or where it sets either the loss or gradients or LR temporarily to 0, thereby changing no parameters at all and effectively skipping that step.

**Comment by gwern** [comment]
shawwn has experimented with just stop loss using unbalanced BigGANs. We can't see any particular worsening of stability or loss curves, and it feels somewhat better and is simpler. It appears the gradient ascent part is not particularly necessary for GANs, at least for stabilizing training early on. (There may still be an argument for doing the gradient ascent somewhere around convergence, where it's more comparable to a converged CNN.)

**Comment by gwern** [comment]
Still no particular observations suggesting the gradient ascent part of flood loss is necessary for GANs: stop loss continues to work.

Worth noting: shawwn has logged individual iterations (rather than the usual averaged TPU pod iterations you see in Tensorboard) to look at dynamics. The loss patterns per iteration are quite odd and there seem to be very rapid oscillations which are masked by averaging: https://twitter.com/theshawwn/status/1261510638292971520 It may be that the usual hinge loss + `D_iter=2` BigGAN training may be a bad idea, at least with stop-loss. 

2 steps per G steps is justified in the BigGAN paper as simply the best setting found out of the range 1-6, but that didn't benefit from the stabilizing effect of stop-loss. Possibly with stop-loss, only 1 D step per G step is necessary, which would spend less compute on D training and also simplify the code.

**Comment by gwern** [comment]
I think we can make stop loss a default without any further investigation, so I will close this issue as a success.
